Group output support

jokva opened the pull request:
A small patchset  that cleans up some clunky behaviour, a bit of refactoring, and support for some of the G\* keywords (those used by Norne).

jokva committed:
RAII cleanup of test files; WRITE removed

The WRITE test was a mere convenience to create and inspect a summary
file, but this should instead be moved to applications or examples, and
out of tests. The helper class setup now has a destructor that will
automatically attempt to clean up the files produced by ert.

jokva committed:
Support for SUMMARY configured output G-keywords

Add support for a set of G-family summary keywords; productions, rates
and ratios.

jokva committed:
Well dispatcher use partial applications

To clarify intent and the high-level formulas used to calculate the
requested values, a series of partial applications are introduced since
most parameters to the functions are fixed from the caller.

This increases signal-to-noise ratio in the switch dispatch, at the
expense of some boilerplate before the switch itself.

joakim-hove commented on opm/output/eclipse/Summary.cpp:
```diff
@@ -432,6 +458,93 @@ inline double well_keywords( const smspec_node_type* node,
+        case rt::gas: return convert( sum( wells, phase ),
+                                      dim::gas_surface_volume,
+                                      conversion_table );
+        default: return sum( wells, phase );
```
Who goes in the `default:`?

jokva replied:
Ideally nothing, and it could be a throw. The fall-back here is just not applying any unit conversions. Preferences?

joakim-hove commented on opm/output/eclipse/Summary.cpp:
```diff
@@ -432,6 +458,93 @@ inline double well_keywords( const smspec_node_type* node,
+        case E::GGIT: return accu - vol( rt::gas );
+
+        /* Production ratios */
+        case E::GWCT: return rate( rt::wat ) /
```
Potential division by zero error.

jokva committed:
Rename wwct/wgor to wct/gor

These functions don't care if the source is a well, group or region, so
they've been renamed to emphasise its generality and reusability.

jokva committed:
Replace silent wrong default with throw

Not asking sum_rate/vol for wat/oil/gas is a programmer error in the
library and shouldn't happen, but is unfortunately unverifiable with the
current tooling. The fall-through now signals a runtime error instead of
silently returning the wrong value.

jokva committed:
Unsupported keywords write negative values

This provides easier-to-identify behaviour than simply writing zero and
going on its merry way.

jokva committed:
Precompute string-to-enum mapping of keywords

By precomputing and storing the keyword-to-enum mapping, accessing and
writing to a node overhead changes from log(N) to constant, and should
behave better w.r.t. cache and memory access.

jokva commented:
Depends on https://github.com/OPM/opm-parser/pull/779

jokva commented:
The PR has been amended with the following changes:
- Keyword-to-enum mapping is now done once instead of per keyword per timestep. Saves us some computation and opens up for some more interface cleanup later.
- Behaviour in error paths changed - things that **should not happen** unless there is a bug in opm-output throws, and unsupported keyword lookup returns -1.
